src/search/parser/lexical_analyzer.cc

```diff
@@ -0,0 +1,131 @@
+#include "lexical_analyzer.h"
+
+#include "errors.h"
+
+#include "../utils/collections.h"
+#include "../utils/system.h"
+
+#include <cassert>
+#include <regex>
+
+using namespace std;
+
+namespace parser {
+static regex build_regex(const string &s, regex::flag_type f = regex_constants::ECMAScript) {
+    return regex("^\\s*(" + s + ")\\s*", f);
+}
+
+static const vector<pair<TokenType, regex>> token_type_expressions = {
+    {TokenType::OPENING_PARENTHESIS, build_regex("\\(")},
+    {TokenType::CLOSING_PARENTHESIS, build_regex("\\)")},
+    {TokenType::OPENING_BRACKET, build_regex("\\[")},
+    {TokenType::CLOSING_BRACKET, build_regex("\\]")},
+    {TokenType::COMMA, build_regex(",")},
+    {TokenType::EQUALS, build_regex("=")},
+    {TokenType::FLOAT, build_regex(
+       R"([+-]?((\d*\.\d+|\d+\.)(e[+-]?\d+|[kmg])?)\b)",
```

Comment:
Would this work then:
`[+-]?(((\d*\.\d+|\d+\.)(e[+-]?\d+|[kmg]\b)?)|\d+e-\d+)` (https://regex101.com/r/JIQCCJ/1)

To answer the first question, we came up with the expression ourselves, it is not taken from somewhere.

Reply:
If we no longer treat `1e3` as an integer token, the last part of the expression should use `[+-]?` instead of `-`.

Reply:
Makes sense to me. It is a bit combination-of-cases-distinction-y, presumably to avoid accidentally parsing an int as a float. Something like this is commonly addressed by priorities: we first match against an integer regex, and for the float regex it then doesn't matter if it also would match integers. (For example, this is very commonly used to match keywords first and then have general rules for matching identifiers.) I think this could be simpler. For example like this:
`[+-]?(\d*\.\d+|\d+\.?)(e[+-]?\d+|[kmg]\b)?`. (Perhaps you also want/need surround parens for everything for the matching.)

Reply:
Exactly, we don't want to parse an int as a float. In case of integer arguments this causes problems because we cannot convert floats to integer currently (only the other way around).

If we give priority to integer tokens, this leads to another issue. When splitting the string into tokens, we do this by greedily matching our expression to the start of the string. A string like `42.0` would then be parsed as an integer token `42` followed by a float token `.0` but this is not what we want.